augur/io/metadata.py

```diff
@@ -454,7 +454,7 @@ def write_records_to_tsv(records, output_file):
             tsv_writer.writerow(record)
 
 
-def _get_delimiter(path: str, valid_delimiters: Iterable[str]):
+def _get_delimiter(path: str, valid_delimiters: str):
```

Comment:
I had thought about this more but didn't write it down in the commit message.

`DEFAULT_DELIMITERS`, which is clearly a `tuple[Literal[','], Literal['\t']]` type, works:

1. During run-time because `csv.Sniffer.sniff`'s [source code](https://github.com/python/cpython/blob/eaa1eae55ea66d74c5303924320185dac74d4eb1/Lib/csv.py#L354) is written in a way that works for any iterable of single-character strings.
2. During type-checking because the `delimiters` parameter to `read_metadata()` is untyped.
    - Since much of the codebase is untyped, Mypy is not configured to run with the [stricter options](https://mypy.readthedocs.io/en/stable/config_file.html#disallow-dynamic-typing) such as `disallow_any_expr` that would enforce typing of `delimiters`.

The type for the `delimiters` parameter of `csv.Sniffer.sniff()` (defined [here](https://github.com/python/typeshed/blob/b6a422112b0f3a8d621600012871373e7fdb2ee0/stdlib/csv.pyi#L154)) are stricter than run-time. I think this is justified: ideally the type would be something like `Iterable[char]`, however `char` is not a type, so `str` is the most accurate representation of that. During run-time, you can use an iterable of multi-char strings such as `csv.Sniffer().sniff("a,,b,,c", delimiters=[',,'])`, but it results in a generic error. Static type-checking catches this earlier with a better error message.

Instead of relaxing the type for `valid_delimiters`, it should be kept as `str` to ensure only single-character delimiters are used.

There must be plenty of similar inconsistencies. I think the right thing to do here is to add typing incrementally. I'll add another commit that properly types `delimiters`.

I'll also split up 04e70278f6b64caa06b55dc112dd6a5c8e6d55fe to give more explanation for non-trivial changes like this one.

Reply:
Ah, thanks for the additional background. Totally makes sense, especially after seeing the `[',,']` multi-char string example. However, I can see the `str` type tripping people up if they don't think of it as `Iterable[char]`. Inline comments will definitely help this confusion.  

Just thinking out loud if we want to define a custom `IterableChar` type? It could even just be a type alias for `str`.

Reply:
> Just thinking out loud if we want to define a custom `IterableChar` type?

I tried this but it gets ugly if we want to change `DEFAULT_DELIMITERS` to be a string, since currently all the subcommands that implement `--metadata-delimiters` accept a list of strings.

In 8c19b7581b631c83338d32b12a347fb7f7982544, I went with an approach that, at the lowest level possible:

1. Validates each string is a single character
2. Joins the characters together to satisfy the type limitation for `csv.Sniffer.sniff(delimiters)`

Reply:
Thanks for humoring me! Your approach is definitely more straight-forward.